### PR TITLE
fixes #3859 clipnorm tensorflow

### LINKED ISSUE #3859: clipnorm doesn't work with Embedding
I'm getting a Traceback every time "clipnorm" is used in NN with Embedding layer.
Here is a simple script where the problem is obvious:

``` python
import numpy as np
from keras.layers import Input, Embedding
from keras.optimizers import Adam
from keras.models import Model

input_layer = Input(shape = (1,) )

embedding = Embedding(input_dim = 1,
                      output_dim = 1)(input_layer)

model = Model(input = input_layer, output = embedding)

model.compile(optimizer = Adam(clipnorm = 1.0), loss = 'mse')

X = np.array([[1]])
Y = np.array([[[0.5]]])
model.fit(X, Y, nb_epoch = 1)
```

Failure:

``` shell
I tensorflow/core/common_runtime/gpu/gpu_device.cc:867] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:01:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:867] Creating TensorFlow device (/gpu:1) -> (device: 1, name: GeForce GTX TITAN X, pci bus id: 0000:02:00.0)
Traceback (most recent call last):
  File "./clipnorm-bug.py", line 20, in <module>
    model.fit(X, Y, nb_epoch = 1)
  File "/usr/local/lib/python3.5/dist-packages/keras/engine/training.py", line 1079, in fit
    self._make_train_function()
  File "/usr/local/lib/python3.5/dist-packages/keras/engine/training.py", line 696, in _make_train_function
    self.total_loss)
  File "/usr/local/lib/python3.5/dist-packages/keras/optimizers.py", line 379, in get_updates
    grads = self.get_gradients(loss, params)
  File "/usr/local/lib/python3.5/dist-packages/keras/optimizers.py", line 71, in get_gradients
    grads = [clip_norm(g, self.clipnorm, norm) for g in grads]
  File "/usr/local/lib/python3.5/dist-packages/keras/optimizers.py", line 71, in <listcomp>
    grads = [clip_norm(g, self.clipnorm, norm) for g in grads]
  File "/usr/local/lib/python3.5/dist-packages/keras/optimizers.py", line 9, in clip_norm
    g = K.switch(n >= c, g * c / n, g)
TypeError: unsupported operand type(s) for *: 'IndexedSlices' and 'float'
```

Keras version is 1.1.0, TensorFlow is 0.10rc

clipvalue on the other hand works fine.

### PR DESCRIPTION
**Pull request by chenb67**
@kracwarlock code from PR #4915
it works, what needs to be done to get it merged?

### CODE REVIEW

File: keras/optimizers.py [hunk 1]
```diff
@@ -12,7 +12,28 @@
 
 def clip_norm(g, c, n):
     if c > 0:
-        g = K.switch(n >= c, g * c / n, g)
+        if K.backend() == 'tensorflow':
```
Comment (fchollet) [inline keras/optimizers.py]: Add a comment explaining why we need to add shape inference code in the TF case.

File: keras/optimizers.py [hunk 2]
```diff
@@ -11,8 +12,28 @@
...
-    if c > 0:
-        g = K.switch(n >= c, g * c / n, g)
+    if c <= 0:
+        return g
+    if K.backend() == 'tensorflow':
+        condition = n >= c
+        then_expression = tf.scalar_mul(c / n, g)
+        else_expression = g
+
+        if isinstance(then_expression, tf.Tensor):  # saving the shape to avoid converting sparse tensor to dense
```
Comment (fchollet) [inline keras/optimizers.py]: Line too long, put the comment below

File: keras/optimizers.py [hunk 3]
```diff
@@ -11,8 +12,28 @@
 
 
 def clip_norm(g, c, n):
-    if c > 0:
-        g = K.switch(n >= c, g * c / n, g)
+    if c <= 0:
+        return g
+    if K.backend() == 'tensorflow':
```
Comment (fchollet) [inline keras/optimizers.py]: Add a comment explaining why this block is necessary

Comment (chenb67) [inline keras/optimizers.py]: I wasn't sure which block you meant so I added a comment to both..
I hope it's clear now, let me know

**Comment by fchollet** [comment]
- The PR now contains unrelated code, please rebase to master.
- Add a unit test, with complete test coverage of the code added. The test should be failing with master, in order to demonstrate the issue that is being fixed, and should pass with the PR.

**Comment by chenb67** [comment]
sorry for the mess with commits..
also, I wasn't sure where to add the unittest - it involves optimizers and layers so I put it in integration.
let me know if it is ok like this

File: tests/integration_tests/test_temporal_data_tasks.py [hunk 4]
```diff
@@ -204,5 +205,14 @@ def test_masked_temporal():
...
     assert(np.abs(history.history['loss'][-1] - ground_truth) < 0.06)
 
+
+@pytest.mark.skipif(K.backend() != 'tensorflow', reason='Requires TF backend')
+@keras_test
+def test_embedding_with_clipnorm():
+    model = Sequential()
+    model.add(layers.Embedding(input_dim=1, output_dim=1))
+    model.compile(optimizer=optimizers.SGD(clipnorm=0.1), loss='mse')
+    model.fit(np.array([[0]]), np.array([[[0.5]]]), nb_epoch=1)
```
Comment (fchollet) [inline tests/integration_tests/test_temporal_data_tasks.py]: `epochs`

**Comment by fchollet** [comment]
LGTM

**Pull request merged**